### PR TITLE
BlackVue Video Support

### PR DESCRIPTION
**Pull request by stilldavid**
Initial support for BlackVue videos. Will need more testing as more videos come up, but currently works on individual videos and a directory with a sequence of videos with no cuts - directories with cuts currently have undefined behavior.

Adds `pymp4` as a dependency.

### CODE REVIEW

**Comment by ybkuang** [comment]
@stilldavid can you help add pymp4 to the setup as see if the CI pass or not? Thanks!

File: mapillary_tools/process_video.py [hunk 1]
```diff
@@ -35,6 +37,7 @@ def timestamps_from_filename(full_image_list,
 
 def sample_video(video_file,
                  import_path,
+                 geotag_source,
```
Comment (jernejaMislej) [inline mapillary_tools/process_video.py]: required arg geotag_source might break calls for this func , sample_video command in tools does not have this arg

File: mapillary_tools/process_geotag_properties.py [hunk 2]
```diff
@@ -41,8 +41,7 @@ def process_geotag_properties(import_path,
                                                   "failed",
                                                   verbose)
         return
-
-    elif geotag_source != "exif" and not os.path.isfile(geotag_source_path):
+    elif geotag_source not in ["exif", 'blackvue'] and not os.path.isfile(geotag_source_path):
```
Comment (jernejaMislej) [inline mapillary_tools/process_geotag_properties.py]: did you change this just because blackvue geotag source is a dir and not file?
some check is needed, since if geo data is not in exif than it must be somewhere else and that location must exist

File: mapillary_tools/process_video.py [hunk 3]
```diff
@@ -49,30 +52,77 @@ def sample_video(video_file,
 
     # command specific checks
     video_file = os.path.abspath(video_file) if video_file else None
-    if video_file and not os.path.isfile(video_file):
-        print("Error, video file " + video_file +
-              " does not exist, exiting...")
+    if not video_file:
```
Comment (jernejaMislej) [inline mapillary_tools/process_video.py]: same comment as above, was this check simplified just because video_file can be a dir of videos?

File: mapillary_tools/uploader.py [hunk 4]
```diff
@@ -166,6 +166,15 @@ def get_upload_file_list(import_path, skip_subfolders=False):
     return sorted(upload_file_list)
 
 
+# get a list of video files in an import path
+def get_video_file_list(import_path):
+    video_file_list = []
+    for root, dir, files in os.walk(import_path):
+        video_file_list.extend(os.path.join(root, file) for file in files if file.lower().endswith(('mp4')) and preform_upload(
+            root, file))
```
Comment (jernejaMislej) [inline mapillary_tools/uploader.py]: the preform_upload function might not be applicable to video files as the logging logic is different there, why is that added? just copy paste from other listing functions?

File: mapillary_tools/gpx_from_blackvue.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    fd.seek(0)
+
+    while fd.tell() < eof:
+        try:
+            box = Box.parse_stream(fd)
+        except RangeError:
+            print 'error parsing blackvue GPS information, exiting'
+            sys.exit()
+        except ConstError:
+            print 'error parsing blackvue GPS information, exiting'
```
Comment (jernejaMislej) [inline mapillary_tools/gpx_from_blackvue.py]: print

File: mapillary_tools/process_video.py [hunk 6]
```diff
@@ -49,30 +52,77 @@ def sample_video(video_file,
...
+def extract_frames(video_file,
+                   import_path,
+                   video_sample_interval,
+                   video_start_time=None,
+                   video_duration_ratio=1.0,
+                   verbose=False,
+                   start_number=None):
+
+    if verbose:
+        print 'extracting frames from', video_file
```
Comment (jernejaMislej) [inline mapillary_tools/process_video.py]: print

File: mapillary_tools/processing.py [hunk 7]
```diff
@@ -210,7 +241,7 @@ def geotag_from_gps_trace(process_file_list,
 
     if use_gps_start_time:
         # update offset time with the gps start time
-        offset_time += (sub_second_times[0] - gps_trace[0][0]).total_seconds()
+        offset_time += (sorted(sub_second_times)[0] - gps_trace[0][0]).total_seconds()
```
Comment (jernejaMislej) [inline mapillary_tools/processing.py]: yes i think you r right, this should be sorted to take the smallest time, i will add it to master already since it could result in jumps otherwise

File: requirements.txt [hunk 8]
```diff
@@ -1,2 +1,3 @@
 -e .
-git+https://github.com/mapillary/Piexif
\ No newline at end of file
+git+https://github.com/mapillary/Piexif
+pymp4==1.1.0
```
Comment (jernejaMislej) [inline requirements.txt]: not needed here, since it is in setup.py
